JitArm64: Implement CR cache

MerryMage opened the pull request:
* Here we mix allocation of registers to guest GPRs and CRs.
* GenMfcr doesn't work, removed it.
* Possible alternative: Allocating a subset of host registers to guest CRs.

lioncash commented on Source/Core/Core/PowerPC/JitArm64/JitArm64_RegCache.h:
```diff
@@ -216,16 +217,19 @@ class Arm64GPRCache : public Arm64RegCache
+  u32 GetImm(size_t preg) const { return GetGuestGprOpArg(preg).GetImm(); }
+
+  void BindToRegister(size_t preg, bool do_load) { BindToRegister(GetGuestGpr(preg), do_load); }
+  void BindCrToRegister(size_t preg, bool do_load) { BindToRegister(GetGuestCr(preg), do_load); }
```
Initialisms/abbreviations should be capitalized.

lioncash commented on Source/Core/Core/PowerPC/JitArm64/JitArm64_RegCache.cpp:
```diff
@@ -105,35 +115,71 @@ bool Arm64GPRCache::IsCalleeSaved(ARM64Reg reg)
+const OpArg& Arm64GPRCache::GetGuestGprOpArg(size_t preg) const
 {
-  OpArg& reg = m_guest_registers[preg];
+  _assert_(preg < 32);
```
You may want to make a constexpr constant to make this less of a magic value.

lioncash commented on Source/Core/Core/PowerPC/JitArm64/JitArm64_RegCache.cpp:
```diff
@@ -105,35 +115,63 @@ bool Arm64GPRCache::IsCalleeSaved(ARM64Reg reg)
+
+Arm64GPRCache::GuestRegInfo Arm64GPRCache::GetGuestCr(size_t preg)
+{
+  _assert_(preg < 8);
```
Ditto.

lioncash commented on Source/Core/Core/PowerPC/JitArm64/JitArm64_RegCache.h:
```diff
@@ -216,17 +217,16 @@ class Arm64GPRCache : public Arm64RegCache
-  ARM64Reg R(size_t preg);
-
+  ARM64Reg R(size_t preg) { return R(GetGuestGpr(preg)); }
+  ARM64Reg Cr(size_t preg) { return R(GetGuestCr(preg)); }
```
Initialisms should be uppercased.

lioncash commented on Source/Core/Core/PowerPC/JitArm64/JitArm64_RegCache.h:
```diff
@@ -236,15 +236,31 @@ class Arm64GPRCache : public Arm64RegCache
+    OpArg& reg;
+  };
+
+  const OpArg& GetGuestGprOpArg(size_t preg) const;
```
Ditto here.

degasus commented on Source/Core/Core/PowerPC/JitArm64/JitArm64_Integer.cpp:
```diff
@@ -425,17 +415,17 @@ void JitArm64::cmp(UGeckoInstruction inst)
-
-  SUB(XA, XA, XB);
-  STR(INDEX_UNSIGNED, XA, PPC_REG, PPCSTATE_OFF(cr_val[crf]));
+  SUB(gpr.Cr(crf), XA, XB);
```
I doubt WB is needed here. With a better emitter and SUB+SXTW in one instruction, XA isn't needed either.

degasus commented on Source/Core/Core/PowerPC/JitArm64/JitArm64_SystemRegisters.cpp:
```diff
@@ -107,8 +93,7 @@ void JitArm64::mcrxr(UGeckoInstruction inst)
   MOVP2R(XB, m_crTable.data());
-  LDR(XB, XB, XA);
-  STR(INDEX_UNSIGNED, XB, PPC_REG, PPCSTATE_OFF(cr_val[inst.CRFD]));
+  LDR(gpr.Cr(inst.CRFD), XB, XA);
```
WB isn't needed here.

degasus commented on Source/Core/Core/PowerPC/JitArm64/JitArm64_SystemRegisters.cpp:
```diff
@@ -585,7 +561,7 @@ void JitArm64::crXXX(UGeckoInstruction inst)
   int bit = 3 - (inst.CRBD & 3);
 
-  LDR(INDEX_UNSIGNED, XB, PPC_REG, PPCSTATE_OFF(cr_val[field]));
+  MOV(XB, gpr.Cr(field));
```
Unlock(WB); gpr.BindCrToRegister(field, true); XB = gpr.Cr(field);

degasus commented on Source/Core/Core/PowerPC/JitArm64/JitArm64_SystemRegisters.cpp:
```diff
@@ -634,9 +612,40 @@ void JitArm64::mfcr(UGeckoInstruction inst)
-  gpr.Lock(W0, W1, W2, W30);
-  BL(GetAsmRoutines()->mfcr);
-  gpr.Unlock(W1, W2, W30);
+  gpr.Lock(W0, W1);
```
yeah, nope. No comment :P

degasus commented on Source/Core/Core/PowerPC/JitArm64/JitArm64_SystemRegisters.cpp:
```diff
@@ -672,8 +681,8 @@ void JitArm64::mtcrf(UGeckoInstruction inst)
-        LDR(XA, XB, ArithOption(XA, true));
-        STR(INDEX_UNSIGNED, XA, PPC_REG, PPCSTATE_OFF(cr_val[i]));
+        gpr.BindCrToRegister(i, false);
+        LDR(gpr.Cr(i), XB, ArithOption(XA, true));
```
WA isn't needed any more.

degasus commented on Source/Core/Core/PowerPC/JitArm64/JitAsm.cpp:
```diff
@@ -651,47 +651,10 @@ void JitArm64::GenerateCommonAsm()
+  GetAsmRoutines()->mfcr = nullptr;
 }
 
 void JitArm64::GenMfcr()
```
Feel free to just drop this function

degasus commented on Source/Core/Core/PowerPC/JitArm64/JitArm64_RegCache.h:
```diff
@@ -115,6 +115,7 @@ class HostReg
   bool m_locked;
 };
 
+template <size_t GUEST_REG_COUNT>
```
I wonder why you've choosen a template instead of a runtime parameter. std::vector isn't that slow ;)

But I don't care

degasus commented on Source/Core/Core/PowerPC/JitArm64/JitArm64_RegCache.cpp:
```diff
@@ -143,26 +182,40 @@ void Arm64GPRCache::FlushRegister(size_t preg, bool maintain_state)
+
   for (size_t i = 0; i < m_guest_registers.size(); ++i)
   {
     if (regs[i])
```
Isn't regs[i] out of bounds?

MerryMage replied:
Good catch. Will have to update to use BitSet64 since we have more that 32 guest registers now.

degasus replied:
I think it is also fine to flush all CR regs here. This bitfield is AFAIK only used to not flush a temporary register within load-store instructions.

MerryMage commented:
Changes:

* Use a std::vector for m_guest_registers instead of using templates.
* Capitalised initialisms
* Removed GenMfcr
* Removed unnecessary scratch registers

degasus commented on Source/Core/Core/PowerPC/JitArm64/JitArm64_SystemRegisters.cpp:
```diff
@@ -623,25 +602,59 @@ void JitArm64::crXXX(UGeckoInstruction inst)
+
+  for (int i = 0; i < 8; i++)
+  {
+    MOV(XB, gpr.CR(i));
```
redundant mov

degasus commented on Source/Core/Core/PowerPC/JitArm64/JitArm64_SystemRegisters.cpp:
```diff
@@ -623,25 +602,55 @@ void JitArm64::crXXX(UGeckoInstruction inst)
+    {
+      ORR(WA, WZR, WA, ArithOption(WA, ST_LSL, 4));
+      UBFX(XC, XB, 61, 1);
+      ORR(XA, XA, XC);
```
As the upper bits of WA are zero, this is the same as:
UBFX(XC, XB, 61, 1);
ORR(WA, XC, XA, ArithOption(XA, ST_LSL, 4));

degasus commented on Source/Core/Core/PowerPC/JitArm64/JitArm64_SystemRegisters.cpp:
```diff
@@ -653,29 +662,30 @@ void JitArm64::mtcrf(UGeckoInstruction inst)
     MOVP2R(XB, m_crTable.data());
     for (int i = 0; i < 8; ++i)
     {
+      gpr.BindCRToRegister(i, false);
```
CR is only written if (crm & (0x80 >> i)) != 0

degasus commented:
LGTM

lioncash commented on Source/Core/Core/PowerPC/JitArm64/JitArm64_RegCache.h:
```diff
@@ -118,7 +117,9 @@ class HostReg
 {
 public:
-  Arm64RegCache() : m_emit(nullptr), m_float_emit(nullptr), m_reg_stats(nullptr){};
+  Arm64RegCache(size_t guest_reg_count)
```
This should be `explicit`
